### ISSUE TITLE
Arity-1 name-based extractors should work like arity 2+

### ISSUE THREAD

**Issue by scabug** [scala/bug#9836]
```scala
Welcome to Scala 2.11.8 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_102).
Type in expressions for evaluation. Or try :help.

scala> :pa
// Entering paste mode (ctrl-D to finish)

final class Foo1(a: Int) {
  def isEmpty = false
  def get     = this
  def _1      = a
}
object Foo1 { def unapply(x: Foo1) = x }

final class Foo2(a: Int, b: Int) {
  def isEmpty = false
  def get     = this
  def _1      = a
  def _2      = b
}
object Foo2 { def unapply(x: Foo2) = x }


// Exiting paste mode, now interpreting.

defined class Foo1
defined object Foo1
defined class Foo2
defined object Foo2

scala> new Foo1(1) match { case Foo1(a) => Array(a) }
res0: Array[Foo1] = Array(Foo1@6340e5f0)

scala> new Foo2(1, 2) match { case Foo2(a, b) => Array(a, b) }
res1: Array[Int] = Array(1, 2)
```

For clarity, I expected `res0` to be `Array[Int]` not `Array[Foo1]`.

I believe this is as specced, but I submit as counter-proof that @adriaanm expected it to work as well: "Looks like it doesn't work for Product1":

https://issues.scala-lang.org/browse/SI-8989?focusedCommentId=71182

**Comment by dwijnand** [comment]
Tested on Dotty and the same problem exists there: https://scastie.scala-lang.org/kVbcuVRFQ7ScUjx5VehItA.

**Comment by som-snytt** [comment]
I started looking at this and the linked tickets about so-called "auto-tupling in patterns".

I think the first example below is perfectly fine and should not warn (as it currently does). I almost had a fix last week, but `ExtractorAlignment` sees the tree in typer and the different tree in patmat. The test is simply that the subpattern consumes the required tuple.

The code is commented "don't unwrap Tuple1", but maybe there is a useful cue for whether to unwrap, for example, `Product1`, or detect `get = this` to avoid self-extraction. I encountered the asymmetry in a macro once, where it was inconvenient to generate the special case.

```
object X { def unapply(s: String): Option[(Int,Int)] = { val Array(a,b) = s.split("/") ; Some((a.toInt,b.toInt)) }}
object T { def unapply(t: (Int,Int)) = Option(t._1+t._2) }

trait Test {
  "1/2" match { case X(T(x)) => x }

  "1/3" match { case X(x @ (_, _)) => x }

  "1/4" match { case X(x) => x }
}
```
